base_layer/core/src/proof_of_work/lwma_diff.rs

```diff
@@ -29,27 +29,29 @@ impl LinearWeightedMovingAverage {
         Self {
             target_difficulties: Vec::with_capacity(block_window + 1),
             block_window,
-            target_time,
+            target_time: target_time as u128,
             max_block_time,
         }
     }
 
     fn calculate(&self) -> Option<Difficulty> {
+        // This function uses u128 internally for most of the math as its possible to have an overflow with large
+        // difficulties and large block windows
         if self.target_difficulties.len() <= 1 {
             return None;
         }
 
         // Use the array length rather than block_window to include early cases where the no. of pts < block_window
-        let n = (self.target_difficulties.len() - 1) as u64;
+        let n = (self.target_difficulties.len() - 1) as u128;
 
-        let mut weighted_times: u64 = 0;
+        let mut weighted_times: u128 = 0;
         let difficulty = self
             .target_difficulties
             .iter()
             .skip(1)
-            .fold(0u64, |difficulty, (_, d)| difficulty + d.as_u64());
+            .fold(0u128, |difficulty, (_, d)| difficulty + d.as_u64() as u128);
 
-        let ave_difficulty = difficulty as f64 / n as f64;
+        let ave_difficulty = difficulty / n as u128;
```

Comment:
Isn't `n` already cast to `u128` on line 45?

Reply:
yip should have removed this as soon as I declared n as u128